src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,175 @@
+        /// Returns the EntitySetName.
+        /// </summary>
+        public string EntitySetName
+        {
+            get
+            {
+                return this.entitySetName;
+            }
+        }
+
+        /// <summary>
+        /// Computes the EntyitySetName in cases where the top-level EntityDescriptors
+        /// do not have the EntitySetName set. 
+        /// </summary>
+        public void ComputeEntitySetName()
+        {
+            EntityDescriptor parentDescriptor = this.TopLevelDescriptors[0] as EntityDescriptor;
+
+            if (!string.IsNullOrEmpty(parentDescriptor.EntitySetName))
+            {
+                this.entitySetName = parentDescriptor.EntitySetName;
+            }
+            else
+            {
+                Uri baseUri = new Uri(this.requestInfo.BaseUriResolver.GetBaseUriWithSlash().ToString());
+                Uri resourceUri = parentDescriptor.EditLink;
+                ODataUriParser parser = new ODataUriParser(this.requestInfo.Format.ServiceModel, baseUri, resourceUri);
+                ODataPath path = parser.ParsePath();
+
+                if (path != null)
+                {
+                    EntitySetSegment entitySetSegment = path.ToList().OfType<EntitySetSegment>().LastOrDefault();
+                    this.entitySetName = entitySetSegment.Identifier;
+                }
+                else
+                {
+                    throw Error.InvalidOperation(Strings.DataBinding_Util_UnknownEntitySetName(parentDescriptor.Entity.GetType().FullName));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Links a descriptor with its related descriptors.
+        /// </summary>
+        /// <param name="parent">The parent descriptor.</param>
+        /// <param name="child">The related descriptor to the parent descriptor.</param>
+        public void AddRelatedDescriptor(Descriptor parent, Descriptor child)
+        {
+            HashSet<Descriptor> childrenDescriptors = this.GetRelatedDescriptors(parent);
+
+            if (childrenDescriptors.Count > 0)
+            {
+                // If the parent already has some related descriptors then we don't need to 
+                // create a new parent-child entry in this.relatedDescriptors dictionary, we just add the child descriptor to the
+                // parent descriptors related collection. 
+                childrenDescriptors.Add(child);
+            }
+            else 
+            {
+                childrenDescriptors.Add(child);
+            }      
+        }
+
+        /// <summary>
+        /// Adds to the top-level descriptors' list.
+        /// </summary>
+        /// <param name="descriptor">The descriptor to be added to the top-level descriptors' list.</param>
+        public void AddTopLevelDescriptors(EntityDescriptor descriptor)
+        {
+            if (this.topLevelDescriptors == null)
+            {
+                this.topLevelDescriptors = new List<Descriptor>();

```

Comment:
I still suggest that you make `topLevelDescriptors` a `List<EntityDescriptor>` since you're not adding any other descriptor kinds to it. It will remove the need to cast the `Descritpor` back to `EntityDescriptor` later when you're processing the list.